### ISSUE TITLE
python vs R for basic plots, tables, and metric calculation

### ISSUE THREAD

**Issue by kbartelt** [Riverscapes/RIM#5]
Hi @philipbaileynar, I was hoping to get your advice on this:

Currently after I manually map DCE shapefiles I am calculating summary metrics and producing some basic tables and plots in R. At this point I need to make a lot of updates to the R script and was talking with @joewheaton about whether it would be advantageous to code this in python instead. 

Joe suggested moving everything from R to python, particularly because i'm not doing anything requiring R and we would avoid a potential user needing to use two different programs. Do you recommend matplotlib for basic summary tables and plots such as bar plots or pie charts? Do you know if this is much more difficult to use than something like ggplot? I think it would be faster for me to keep this in R, but if moving it to python seems like a much better idea then I'd like to do it that way.

**Comment by philipbaileynar** [comment]
I don't know R so I can't speak fairly. 

I will say that MatPlotLib is extremely straightforward. The internet is full of examples that you can cut and paste for pretty much anything you might want to do. Out of the box, MatPlotLib might not produce plots quite as beautiful as R, but it is extremely customizable and you will be able to achieve everything you need with a little effort. 

By default use CSV files for tables. They are portable to other scripting languages and for most folks they double click open in Excel. If you want something prettier then you can write tables to HTML (and include your PNG graphics from MatPlotLib).

Happy to discuss strategies for any of this...

**Comment by philipbaileynar** [comment]
Oh and I've not heard of ggplot.

**Comment by kbartelt** [comment]
Great, thanks for the input on this. Another question I have along these lines is if it is better practice to try to put all my functions into a single large script or if it is better to have several, smaller scripts? For example, I might have a script with functions that creates my project structure, a separate script with functions that creates a new DCE, and then separate scripts for calculating metrics for a single DCE or multiple DCE's

**Comment by philipbaileynar** [comment]
ooooooooooh Karen, this is a big question that will divide developers!

My short answer is "get the job done", meaning put the code in one long script. The reasons to break the code out into separate functions or files are:

* Your code should avoid repeat itself. So if a block of code needs to get run multiple times then break it out into functions.
* If you are building a big piece of software then separate files allow you to reuse just the bits of code you need. (You can't reuse a few lines in the middle of a big long script)
* Functional Testing - I often break standalone blocks of code into their own functions so that I can put a "main" at the bottom of each file, run and test just the code in that file without running the entire system every time. This is extremely useful for big slow models where you want to develop just one small part without running lengthy processes with every change.
* Unit Testing - Professional software is accompanied by [unit testing](https://en.wikipedia.org/wiki/Unit_testing) code that verifies that each function works as intended. The unit testing code performs is not considered part of the product and doesn't ship with the product. Unit testing relies on the code being broken down into separate functions for each "unit" of work.
* Break code into functions and separate files if it will make it easier to comprehend by other developers.

The bottom line is that if you are hacking together a quick script to get a one time job done, then just throw it in a single long script. If you're building reusable software that might grow over time, get re-used, adopted by other developers, needs testing, then break the code up. Just don't spend a ton of time separating code and architecting it beautifully if it's just you running it one time!

Another software term is [Minimum Viable Product](https://en.wikipedia.org/wiki/Minimum_viable_product) (MVP). Do the least amount of work to get the current task done. Then refactor as necessary later.

**Comment by kbartelt** [comment]
Ah okay, I had a feeling there might be some nuance behind that one. Thank you for explaining and breaking it down for me.
